### DISCUSSION TITLE
Is it possible to check for Production or Development mode?

### DISCUSSION THREAD

**Discussion by retifrav** [fastapi/fastapi#11847] [Questions] [answered]
### Description

The `fastapi run` starts FastAPI in Production mode, and `fastapi dev` starts FastAPI in Development mode. But is it possible to detect/check which mode is it running in now? I'd expect something like:

``` py
api = FastAPI()

if api.DevelopmentMode:
    print("Running in Development mode")
else:
    print("Running in Production mode")
```

I couldn't find anything related in the [FastAPI](https://fastapi.tiangolo.com/reference/fastapi/) class reference.

### Operating System

macOS

### FastAPI Version

0.111.0

### Pydantic Version

2.7.3

### Python Version

3.12.3

**Comment by YuriiMotov** [comment, marked as answer]
FastAPI CLI doesn't pass this information to FastAPI app.

The only difference between these commands is that FastAPI CLI uses different defaults for parameters to run ASGI server (`host` address and `reload`).

And since it's only default parameters, they can be overridden. For example command `fastapi run main.py --host 127.0.0.1 --reload` is a full equivalent of `fastapi dev main.py`.

As I know the common way to determine whether the app is running in prod or dev environment is to use ENV variable.
Just set ENV variable with the name like `RUN_MODE` to `DEV` on your local machine and to `PROD` on your production server

```
export RUN_MODE="DEV"
```

Then read this variable in your app:
```
from fastapi import FastAPI
import os

app = FastAPI()

@app.get("/")
def index():
    return os.environ.get('RUN_MODE', 'NONE')
```

**Reply by retifrav** [reply]
Yes, that's exactly how I did it for now - with an environment variable. Just wanted to check for "in-code" options, in case I missed something from the documentation. Thank you for the explanation!

**Comment by Grz3hu** [comment]
You can simply check what command has been used using system library
```from typing import Annotated
from fastapi import FastAPI, Query

app = FastAPI()

import sys
print(str(sys.argv))
```

output:
`['/home/grz3hu/Documents/Projects/Python/FastAPI/MySample/env/bin/fastapi', 'run']`

**Reply by dhdaines** [reply]
Thanks! Indeed, this works quite well.

It is rather common to use an environment variable for this (to send Access-Control-Allow-Origin: http://localhost for example), would be nice if FastAPI CLI supported this pattern directly (since it otherwise tries to DWIM in many ways)

**Reply by sheecegardezi** [reply]
Thanks for sharing this pattern, it will be nice to have a wrapper on top of this buildin fastapi .
